plat/common/lcpu.c

```diff
@@ -48,3 +165,424 @@ void ukplat_lcpu_halt_to(__nsec until)
+	 * It is safe to do it afterwards, because the RUN IRQ handler will
+	 * only take one function and return afterwards. And we only raise the
+	 * IRQ after finishing setup.
+	 */
+	lcpu->fn.user = fn->user;
+
+	/* Ensure everything is written back when we return and the arch
+	 * support code will raise the IRQ
+	 */
+	wmb();
+
+	return 0;
+}
+
+static void lcpu_fn_dequeue(struct lcpu *this_lcpu, struct ukplat_lcpu_func *fn)
+{
+	*fn = this_lcpu->fn;
+
+	/* Ensure that we have captured the whole function object */
+	rmb();
+
+	UK_ASSERT(fn->fn);
+
+	/* Free the slot. Another function object can be queued afterwards */
+	this_lcpu->fn.fn = NULL;
+}
+
+static int lcpu_ipi_run_handler(void *args __unused, struct __regs *regs)
+{
+	struct lcpu *this_lcpu = lcpu_get_current();
+	struct ukplat_lcpu_func fn;
+
+	lcpu_fn_dequeue(this_lcpu, &fn);
+	fn.fn(regs, &fn);
+
+	/* If we had a transition from BUSY to HALTED in fn, we would
+	 * not reach this code but sit in the error halt loop. We can
+	 * thus safely just decrement without worrying about the HALTED
+	 * state.
+	 */
+	UK_ASSERT(lcpu_state_is_busy(this_lcpu->state));
+	ukarch_dec(&this_lcpu->state);
+
+	return 1;
+}
+
+static int lcpu_ipi_wakeup_handler(void *args __unused,
+				   struct __regs *regs __unused)
+{
+	/* Nothing to do */
+	return 1;
+}
+
+int lcpu_mp_init(void *arg)
+{
+	int rc;
+
+	/* Make sure this is run on the BSP only */
+	UK_ASSERT(lcpu_count == 1);
+	UK_ASSERT(lcpu_get_current() == lcpu_get(0));
+
+	/* Initialize architecture-dependent functionality. This will also do
+	 * CPU discovery and allocation
+	 */
+	rc = lcpu_arch_mp_init(arg);
+	if (unlikely(rc))
+		return rc;
+
+	/* Register the lcpu_run and lcpu_wakeup interrupt handlers */
+	rc = ukplat_irq_register(LCPU_ARCH_IPI_RUN_IRQ,
+				 lcpu_ipi_run_handler, NULL);
+	if (unlikely(rc)) {
+		uk_pr_crit("Could not register handler for IPI IRQ %d\n",
+			   LCPU_ARCH_IPI_RUN_IRQ);
+		return rc;
+	}
+
+	rc = ukplat_irq_register(LCPU_ARCH_IPI_WAKEUP_IRQ,
+				 lcpu_ipi_wakeup_handler, NULL);
+	if (unlikely(rc)) {
+		uk_pr_crit("Could not register handler for wakeup IRQ %d\n",
+			   LCPU_ARCH_IPI_WAKEUP_IRQ);
+		return rc;
+	}
+
+	return 0;
+}
+
+void __weak __noreturn lcpu_entry_default(struct lcpu *this_lcpu)
+{
+	int rc;
+
+	/* Finish initialization. As there is nothing to return to, we
+	 * just enter halted state if an error occurs.
+	 */
+	rc = lcpu_init(this_lcpu);
+	if (unlikely(rc))
+		lcpu_halt(this_lcpu, rc);
+
+	/* We are coming from BUSY0 state and want to transition to IDLE
+	 * state. However, there can already be functions queued so we have
+	 * to use a decrement here.
+	 */
+	ukarch_dec(&this_lcpu->state);
+
+	/* Enable IRQs. If there are functions queued we will immediately
+	 * jump to the IRQ handler.
+	 */
+	ukplat_lcpu_enable_irq();
+	while (1) {
+		/* Besides interrupts in general, the halt can be
+		 * interrupted by calls to ukplat_lcpu_run().
+		 */
+		halt();
```

Comment:
Same here. Should be `lcpu_halt_to_irq()`, and we should fix the arm implementation of `halt()` to be uninterruptible (currently uses `wfi` which is effectively `lcpu_halt_to_irq()`.

Reply:
I think this is correct. `halt` should just put the CPU into a low-power state where it waits for interrupts. If you want an uninterruptable halt to have to call `ukplat_lcpu_halt`, which disables IRQs first (although it is buggy in that it should also have a loop around `halt` to also remain in this state if there is a non-maskable interrupt. Only a CPU reset should be able to recover the CPU). But that is fixed with this PR.

Reply:
Okay, but then what is the purpose of having an `lcpu_halt_to_irq()` function?

Reply:
The `ukplat_lcpu_halt_irq` is the public interface and `halt` is just the function that implements the actual halt (e.g., `hlt` or `wfi`) like `cpuid` or `readmsr` just call the underlying ASM instructions. But you are right, that it is somewhat confusing and we should call `halt` something like `lcpu_halt_irq`. We can do this as a separate PR as it will touch all platforms. But it makes sense. As `halt` is implemented along `reset` and `system_off`, it is clearly an artifact from designing for uniprocessor systems. It has a pretty global touch to it.

Reply:
Then maybe in that case it would be better if we were using the `ukplat_lcpu_` interface consistently in this scope, rather than using in some parts the higher level interface and in some others the low level implementation?

> We can do this as a separate PR as it will touch all platforms

Okay 👍🏼